src/Functions/FunctionBinaryArithmetic.h

```diff
@@ -1326,6 +1407,22 @@ class FunctionBinaryArithmetic : public IFunction
             return getReturnTypeImplStatic(new_arguments, context);
         }
 
+
+        if (isArray(arguments[0]) && isArray(arguments[1]))
+        {
+            if constexpr (is_plus || is_minus)
+            {
+                DataTypes new_arguments {
+                        static_cast<const DataTypeArray &>(*arguments[0]).getNestedType(),
+                        static_cast<const DataTypeArray &>(*arguments[1]).getNestedType(),
+                };
+                return std::make_shared<DataTypeArray>(getReturnTypeImplStatic(new_arguments, context));
+            }
+            else
+                throw Exception(ErrorCodes::UNSUPPORTED_METHOD, "Cannot use this operation on arrays");
```

Comment:
I think we shouldn't throw a separate error here, but let the `ILLEGAL_TYPE_OF_ARGUMENT` error be thrown on line 1615:
```
        throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal types {} and {} of arguments of function {}",
            arguments[0]->getName(), arguments[1]->getName(), String(name));
```

This would be consistent behavior for users because they would not get different error messages for the same error.

Also, why the `constexpr` condition is inside? As that is done in compilation time, (theoretically) it is better not to execute the `isArray(arguments[0]) && isArray(arguments[1])` check when we are not in a `plus` or `minus` function. In this occasion I think the performance difference is unmeasurable, therefore I wouldn't comment on it, but because of the thrown error this should be modified, I dared to comment also on this to make the code a bit better.